src/oc/change/async/persistence.clj

```diff
@@ -52,33 +52,57 @@
     ; return the differences newly changed items & already seen items
     (vec (clojure.set/difference changed-items seen-items))))
 
+
+(defn- unread-items-for
+  "
+  Implements the unread item logic based on changes in the container, when
+  individual items were read.
+  "
+  [container-id all-changes all-reads]
+  (let [changes (filter #(= container-id (:container-id %)) all-changes) ; only changes for this container-id
+        reads (filter #(= container-id (:container-id %)) all-reads) ; only reads for this container-id
+        changed-items (set (map :item-id changes)) ; item ids of the changes for this container
+        read-items (set (map :item-id reads))] ; item ids read in the container
+    ; return the differences newly changed items & already read items
+    (vec (clojure.set/difference changed-items read-items))))
+
 (defn status-for
   "
   Given a set of changes to items in containers...
 
   Changes:
   {:container-id '1111-1111-1111', :item-id '2222-2222-2222', :change-at '2018-06-10T14:49:50.883Z'}
   {:container-id '1111-1111-1111', :item-id '3333-3333-3333', :change-at '2018-06-11T14:49:50.986Z'}
+  {:container-id '1111-1111-1111', :item-id '6666-6666-6666', :change-at '2018-06-14T14:49:50.107Z'}
   {:container-id '4444-4444-4444', :item-id '5555-5555-5555', :change-at '2018-06-12T14:49:57.107Z'})
 
-  And a set of seen events for the user (where '9999-9999-9999' means they saw everything in the container)...
+  A set of seen events for the user (where '9999-9999-9999' means they saw everything in the container)...
 
   Seens:
   {:container-id '1111-1111-1111', :item-id '2222-2222-2222', :seen-at '2018-06-11T11:23:51.395Z'}
   {:container-id '4444-4444-4444', :item-id '9999-9999-9999', :seen-at '2018-06-13T11:23:51.395Z'}
 
+  A set of read events for the user...
+
+  Reads:
+  {:container-id '1111-1111-1111', :item-id '2222-2222-2222', :read-at '2018-06-11T13:23:51.395Z'}
+  {:container-id '1111-1111-1111', :item-id '3333-3333-3333', :read-at '2018-06-13T15:23:51.395Z'}
```

Comment:
In theory that's possible because being "seen" is a container concept (not tracked per entry) but being "read" is an per entry concept. So it's possible to have 10 new items in the container, you haven't gone to the container (section). But then to go directly to the 5th of those 10 items with a direct link to it. So now you've read that 5th item, but the container still has reports the 10 unseen items. The two concepts are just used for different things. The idea of "seen" is used for bolding containers to draw attention to them. It's not perfect, but closing that gap isn't worth it right now.